modcc/solvers.cpp

```diff
@@ -528,6 +529,258 @@ void LinearSolverVisitor::finalize() {
+        rhs.push_back(symtbl_.define(id));
+    }
+
+    A_.augment(rhs);
+    symge::gj_reduce(A_, symtbl_);
+
+    // Create and assign intermediate variables.
+    // Save gaussian elimination solution statements in S_
+    std::vector<expression_ptr> S_;
+    for (unsigned i = 0; i < symtbl_.size(); ++i) {
+        symge::symbol s = symtbl_[i];
+
+        if (primitive(s)) continue;
+
+        auto expr = as_expression(definition(s));
+        auto local_t_term = make_unique_local_assign(block_scope_, expr.get(), "s_");
+        auto t_ = local_t_term.id->is_identifier()->spelling();
+        symtbl_.name(s, t_);
+
+        statements_.push_back(std::move(local_t_term.local_decl));
+        S_.push_back(std::move(local_t_term.assignment));
+    }
+
+    // Create the statements that update the temporary state variables
+    // (dvar_temp) after a Newton's iteration and save them in U_
+    std::vector<expression_ptr> U_;
+
+    Location loc;
+    for (unsigned i = 0; i < A_.nrow(); ++i) {
+        const symge::sym_row &row = A_[i];
+        unsigned rhs_col = A_.augcol();
+        int lhs_col = -1;
+        for (unsigned r = 0; r < A_.nrow(); r++) {
+            if (row[r]) {
+                lhs_col = r;
+                break;
+            }
+        }
+
+        if (lhs_col == -1) {
+            error({"Something went wrong in solving the sparse matrix", loc});
+            return;
+        }
+
+        auto var = make_expression<IdentifierExpression>(loc, dvar_temp_[lhs_col]);
+        auto expr = make_expression<AssignmentExpression>(loc, var->clone(),
+                        make_expression<SubBinaryExpression>(loc, var->clone(),
+                            make_expression<DivBinaryExpression>(loc, make_expression<IdentifierExpression>(loc, symge::name(A_[i][rhs_col])),
+                                make_expression<IdentifierExpression>(loc, symge::name(A_[i][lhs_col])))));
+
+        U_.push_back(std::move(expr));
+    }
+
+    // Do 3 Newton iterations
+    for (unsigned n = 0; n < 3; n++) {
+        // Print out the statements that calulate F(xn), J(xn), solve J(xn)^-1*F(xn), update xn -> xn+1
+        for (auto &s: F_) {
+            statements_.push_back(s->clone());
+        }
+        for (auto &s: J_) {
+            statements_.push_back(s->clone());
+        }
+        for (auto &s: S_) {
+            statements_.push_back(s->clone());
+        }
+        for (auto &s: U_) {
+            statements_.push_back(s->clone());
+        }
+    }
+
+    // Finally update the state variables, dvars_ = dvar_temp_
+    for (unsigned i = 0; i < dvars_.size(); ++i) {
+        auto expr = make_expression<AssignmentExpression>(loc,
+                        make_expression<IdentifierExpression>(loc, dvars_[i]),
+                            make_expression<IdentifierExpression>(loc, dvar_temp_[i]));
+        statements_.push_back(std::move(expr));
+    }
+
+    BlockRewriterBase::finalize();
+}
+
```

Comment:
Perhaps for a different PR, but what do you think about using an explicit step y₀ ← x(t) + δt G(x(t)) for the first approximation in the Newton process?

Reply:
I don't mind that. I could add it now or we can keep it in mind if we need it later?

Reply:
Let's just keep it in mind — it might be a way of keeping a lower bound on the number of Newton iterations, but it's something to be experimented with in the future IMO.